cluster-cm/src/main/java/com/sequenceiq/cloudbreak/cm/ClouderaManagerParcelDecommissionService.java

```diff
@@ -39,59 +39,64 @@
     @Inject
     private ClouderaManagerPollingServiceProvider clouderaManagerPollingServiceProvider;
 
-    public Map<String, String> getParcelsInStatus(ParcelsResourceApi parcelsResourceApi, String stackName, ParcelStatus parcelStatus) {
+    Map<String, String> getParcelsInStatus(ParcelsResourceApi parcelsResourceApi, String stackName, ParcelStatus parcelStatus) {
         try {
-            return getClouderaManagerParcelsByStatus(parcelsResourceApi, stackName, parcelStatus)
-                    .stream()
+            Map<String, String> parcelResponse = getClouderaManagerParcelsByStatus(parcelsResourceApi, stackName, parcelStatus).stream()
                     .collect(toMap(ApiParcel::getProduct, ApiParcel::getVersion));
+            LOGGER.debug("The following parcels are found in {} status: {}", parcelStatus, parcelResponse);
+            return parcelResponse;
         } catch (ApiException e) {
             LOGGER.info("Unable to fetch the list of activated parcels", e);
             throw new ClouderaManagerOperationFailedException("Unable to fetch the list of activated parcels", e);
         }
     }
 
-    public ParcelOperationStatus deactivateUnusedParcels(ParcelsResourceApi parcelsResourceApi, ParcelResourceApi parcelResourceApi,
-            String stackName, Map<String, ClouderaManagerProduct> cmProducts) {
-        Map<String, String> installedComponents = getParcelsInStatus(parcelsResourceApi, stackName, ParcelStatus.ACTIVATED);
-        Map<String, String> filteredParcels = installedComponents.entrySet().stream()
-                .filter(entry -> !cmProducts.containsKey(entry.getKey()))
+    ParcelOperationStatus deactivateUnusedParcels(ParcelsResourceApi parcelsResourceApi, ParcelResourceApi parcelResourceApi,
+            String stackName, Set<String> usedProducts, Set<ClouderaManagerProduct> productsFromImage) {
+        Map<String, String> activeComponents = getParcelsInStatus(parcelsResourceApi, stackName, ParcelStatus.ACTIVATED);
+        Map<String, String> parcelsToDeactivate = getUnusedParcels(activeComponents, usedProducts, productsFromImage);
+        LOGGER.debug("The following parcels will be deactivated: {}", parcelsToDeactivate);
+        return deactivateParcels(parcelResourceApi, stackName, parcelsToDeactivate);
+    }
+
+    private Map<String, String> getUnusedParcels(Map<String, String> availableParcels, Set<String> usedProducts, Set<ClouderaManagerProduct> parcelsFromImage) {
+        return availableParcels.entrySet().stream()
+                .filter(entry -> !usedProducts.contains(entry.getKey())
+                        && parcelsFromImage.stream().anyMatch(product -> product.getName().equals(entry.getKey())))
```

Comment:
I think it would be more performant if you would map `parcelsFromImage` to a `HashSet<Sttring>` and use contains here. It would do the same logically, so it's not a big issue, I don't expect having too many parcels